src/main/java/org/broadinstitute/hellbender/tools/walkers/bqsr/BaseRecalibratorUprooted.java

```diff
@@ -0,0 +1,564 @@
+     * The "features" list doesn't need to be complete but it must cover the read
+     */
+    public void apply( SAMRecord originalRead, final ReferenceContext ref, List<? extends Feature> features ) {
+        ReadTransformer transform = makeReadTransform();
+        final SAMRecord read = transform.apply(originalRead);
+
+        if( ReadUtils.isEmpty(read) ) { return; } // the whole read was inside the adaptor so skip it
+
+        RecalUtils.parsePlatformForRead(read, RAC);
+        if (!RecalUtils.isColorSpaceConsistent(RAC.SOLID_NOCALL_STRATEGY, read)) { // parse the solid color space and check for color no-calls
+            return; // skip this read completely
+        }
+
+        // We've checked in onTraversalStart() that we have a reference, so ref.get() is safe
+        final int[] isSNP = calculateIsSNP(read, ref, originalRead);
+        final int[] isInsertion = calculateIsIndel(read, EventType.BASE_INSERTION);
+        final int[] isDeletion = calculateIsIndel(read, EventType.BASE_DELETION);
+        final int nErrors = nEvents(isSNP, isInsertion, isDeletion);
+
+        // note for efficiency regions we don't compute the BAQ array unless we actually have
+        // some error to marginalize over.  For ILMN data ~85% of reads have no error
+        final byte[] baqArray = nErrors == 0 ? flatBAQArray(read) : calculateBAQArray(read);
+
+        if( baqArray != null ) { // some reads just can't be BAQ'ed
+            final ReadCovariates covariates = RecalUtils.computeCovariates(read, requestedCovariates);
+            final boolean[] skip = calculateSkipArray(read, features); // skip known sites of variation as well as low quality and non-regular bases
+            final double[] snpErrors = calculateFractionalErrorArray(isSNP, baqArray);
+            final double[] insertionErrors = calculateFractionalErrorArray(isInsertion, baqArray);
+            final double[] deletionErrors = calculateFractionalErrorArray(isDeletion, baqArray);
+
+            // aggregate all of the info into our info object, and update the data
+            final ReadRecalibrationInfo info = new ReadRecalibrationInfo(read, covariates, skip, snpErrors, insertionErrors, deletionErrors);
+            recalibrationEngine.updateDataForRead(info);
+        }
+
+        accumulator++;
+    }
+
+    /**
+     * Compute the number of mutational events across all hasEvent vectors
+     *
+     * Simply the sum of entries in hasEvents
+     *
+     * @param hasEvents a vector a vectors of 0 (no event) and 1 (has event)
+     * @return the total number of events across all hasEvent arrays
+     */
+    protected static int nEvents(final int[]... hasEvents) {
+        int n = 0;
+        for ( final int[] hasEvent : hasEvents ) {
+            n += MathUtils.sum(hasEvent);
+        }
+        return n;
+    }
+
+    private boolean[] calculateSkipArray( final SAMRecord read, final List<? extends Feature> features ) {
```

Comment:
we should standardize on style here.  file has a mix of space and no-space between ( and parameters. I prefer no-space, but defer to Broad engineers on this.